File: src/email_automation_jev/classification_repository.py
from __future__ import annotations

import re
from typing import Any

from supabase import Client

from .models import JevClassification
from .repository import utc_now


def sanitized_classification_error(error: BaseException) -> str:
    message = re.sub(r"Bearer\s+\S+", "Bearer [redacted]", str(error), flags=re.I)
    message = re.sub(
        r"(api[_-]?key|token|secret|password)\s*[=:]\s*[^\s,;]+",
        r"\1=[redacted]",
        message,
        flags=re.I,
    )
    return message[:2000]


def create_classifier_version(database: Client, value: dict[str, Any]) -> dict[str, Any]:
    row = {
        **value,
        "reference_config": value.get("reference_config", {}),
        "benchmark_summary": value.get("benchmark_summary"),
    }
    return database.table("classifier_versions").insert(row).execute().data[0]


def find_classifier_version(database: Client, version: str) -> dict[str, Any] | None:
    response = database.table("classifier_versions").select("*").eq("version", version).maybe_single().execute()
    return response.data


def approve_classifier_version(database: Client, version_id: str, summary: dict[str, Any]) -> dict[str, Any]:
    response = (
        database.table("classifier_versions")
        .update({"status": "approved", "benchmark_summary": summary, "approved_at": utc_now()})
        .eq("id", version_id)
        .eq("status", "draft")
        .execute()
    )
    return response.data[0]


def create_classification_run(database: Client, value: dict[str, Any]) -> dict[str, Any]:
    return database.table("classification_runs").insert(value).execute().data[0]


def refresh_classification_run_counters(database: Client, run_id: str) -> None:
    database.rpc("refresh_classification_run_counters", {"p_run_id": run_id}).execute()


def enqueue_classification_emails(database: Client, run_id: str, email_ids: list[str]) -> int:
    rows = [{"run_id": run_id, "email_id": email_id} for email_id in dict.fromkeys(email_ids)]
    if not rows:
        return 0
    response = database.table("email_classification_results").upsert(rows, on_conflict="run_id,email_id", ignore_duplicates=True).execute()
    refresh_classification_run_counters(database, run_id)
    return len(response.data or [])


def mark_classification_running(database: Client, run_id: str, email_id: str) -> None:
    (
        database.table("email_classification_results")
        .update(
            {
                "status": "running",
                "started_at": utc_now(),
                "attempt_count": 1,
                "error_message": None,
            }
        )
        .eq("run_id", run_id)
        .eq("email_id", email_id)
        .eq("status", "queued")
        .execute()
    )


def save_classification_result(database: Client, run_id: str, email_id: str, result: JevClassification) -> None:
    row = {
        "status": "succeeded",
        "category": result.category,
        "category_decision": result.decision,
        "category_confidence": result.confidence,
        "category_top_probability": result.top_probability,
        "category_probabilities": result.probabilities,
        "next_action": result.action["choice"],
        "action_confidence": result.action["confidence"],
        "action_probabilities": result.action["probabilities"],
        "urgency_score": result.urgency["score"],
        "urgency_confidence": result.urgency["confidence"],
        "urgency_probabilities": result.urgency["probabilities"],
        "draft_probability": result.draft_reply["probability"],
        "should_draft": result.draft_reply["should_draft"],
        "model_returned": result.model,
        "input_tokens": result.input_tokens,
        "error_message": None,
        "classified_at": utc_now(),
    }
    (database.table("email_classification_results").update(row).eq("run_id", run_id).eq("email_id", email_id).eq("status", "running").execute())


def fail_classification_result(database: Client, run_id: str, email_id: str, error: BaseException) -> None:
    row = {
        "status": "failed",
        "error_message": sanitized_classification_error(error),
        "classified_at": utc_now(),
    }
    (database.table("email_classification_results").update(row).eq("run_id", run_id).eq("email_id", email_id).eq("status", "running").execute())


def request_classification_cancellation(database: Client, run_id: str) -> None:
    database.table("classification_runs").update({"cancellation_requested_at": utc_now()}).eq("id", run_id).in_("status", ["queued", "running"]).execute()


def append_human_label_event(database: Client, value: dict[str, Any]) -> dict[str, Any]:
    defaults = {
        "source_key": None,
        "category": None,
        "next_action": None,
        "urgency_level": None,
        "draft_needed": None,
        "reviewer_id": None,
        "reviewer_label": None,
        "notes": "",
        "supersedes_event_id": None,
    }
    return database.table("email_human_label_events").insert({**defaults, **value}).execute().data[0]


def create_classification_review_case(database: Client, value: dict[str, Any]) -> dict[str, Any]:
    return database.table("classification_review_cases").insert({"classification_result_id": None, **value}).execute().data[0]


def save_llm_review_suggestion(database: Client, case_id: str, value: dict[str, Any]) -> dict[str, Any]:
    row = {
        "status": "llm_completed",
        "llm_provider": value["provider"],
        "llm_model": value["model"],
        "llm_prompt_version": value["prompt_version"],
        "llm_suggestion": value["suggestion"],
        "llm_input_tokens": value.get("input_tokens"),
        "llm_output_tokens": value.get("output_tokens"),
        "llm_error_message": None,
        "requested_at": utc_now(),
    }
    response = database.table("classification_review_cases").update(row).eq("id", case_id).in_("status", ["pending", "llm_requested"]).execute()
    return response.data[0]

File: lib/classification-repository.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type { JevClassification } from "./jev-classifier.js";
import type {
  ClassificationRunInput,
  ClassificationRunRow,
} from "./classification-types.js";

type DatabaseClient = SupabaseClient;

function assertNoError(error: { message: string } | null, context: string): void {
  if (error) throw new Error(`${context}: ${error.message}`);
}

export function sanitizedClassificationError(error: unknown): string {
  const message = error instanceof Error ? error.message : String(error);
  return message
    .replace(/Bearer\s+[^\s]+/gi, "Bearer [redacted]")
    .replace(/(api[_-]?key|token|secret|password)\s*[=:]\s*[^\s,;]+/gi, "$1=[redacted]")
    .slice(0, 2_000);
}

export async function createClassificationRun(
  database: DatabaseClient,
  input: ClassificationRunInput,
): Promise<ClassificationRunRow> {
  const { data, error } = await database
    .from("classification_runs")
    .insert(input)
    .select("*")
    .single();
  assertNoError(error, "Could not create classification run");
  return data as ClassificationRunRow;
}

export async function enqueueClassificationEmails(
  database: DatabaseClient,
  runId: string,
  emailIds: string[],
): Promise<number> {
  if (emailIds.length === 0) return 0;
  const rows = [...new Set(emailIds)].map((emailId) => ({ run_id: runId, email_id: emailId }));
  const { data, error } = await database
    .from("email_classifications")
    .upsert(rows, { onConflict: "run_id,email_id", ignoreDuplicates: true })
    .select("id");
  assertNoError(error, "Could not enqueue emails for classification");
  await refreshClassificationRunCounters(database, runId);
  return data?.length || 0;
}

export interface ClassificationBatchOutcome {
  emailId: string;
  result?: JevClassification;
  error?: unknown;
}

export function classificationBatchRows(
  runId: string,
  outcomes: ClassificationBatchOutcome[],
  classifiedAt: string,
): Array<Record<string, unknown>> {
  return outcomes.map((outcome) => {
    if (!outcome.result) {
      return {
        run_id: runId,
        email_id: outcome.emailId,
        status: "failed",
        attempt_count: 1,
        category: null,
        category_decision: null,
        category_confidence: null,
        category_top_probability: null,
        category_probabilities: null,
        next_action: null,
        action_confidence: null,
        action_probabilities: null,
        urgency_score: null,
        urgency_confidence: null,
        urgency_probabilities: null,
        draft_probability: null,
        should_draft: null,
        model_returned: null,
        input_tokens: null,
        error_message: sanitizedClassificationError(outcome.error || "Unknown classification failure"),
        classified_at: classifiedAt,
      };
    }
    const result = outcome.result;
    return {
      run_id: runId,
      email_id: outcome.emailId,
      status: "succeeded",
      attempt_count: 1,
      category: result.category,
      category_decision: result.decision,
      category_confidence: result.confidence,
      category_top_probability: result.top_probability,
      category_probabilities: result.probabilities,
      next_action: result.action.choice,
      action_confidence: result.action.confidence,
      action_probabilities: result.action.probabilities,
      urgency_score: result.urgency.score,
      urgency_confidence: result.urgency.confidence,
      urgency_probabilities: result.urgency.probabilities,
      draft_probability: result.draft_reply.probability,
      should_draft: result.draft_reply.should_draft,
      model_returned: result.model,
      input_tokens: result.input_tokens,
      error_message: null,
      classified_at: classifiedAt,
    };
  });
}

export async function saveClassificationBatch(
  database: DatabaseClient,
  runId: string,
  outcomes: ClassificationBatchOutcome[],
): Promise<void> {
  if (outcomes.length === 0) return;
  const classifiedAt = new Date().toISOString();
  const rows = classificationBatchRows(runId, outcomes, classifiedAt);
  const { error } = await database
    .from("email_classifications")
    .upsert(rows, { onConflict: "run_id,email_id", ignoreDuplicates: false });
  assertNoError(error, "Could not save classification batch");

  const latestFailure = [...outcomes].reverse().find((outcome) => !outcome.result);
  if (latestFailure) {
    const runResult = await database
      .from("classification_runs")
      .update({ error_message: sanitizedClassificationError(latestFailure.error) })
      .eq("id", runId)
      .eq("status", "running");
    assertNoError(runResult.error, "Could not record the latest classification error");
  }
}

export async function refreshClassificationRunCounters(
  database: DatabaseClient,
  runId: string,
): Promise<void> {
  const { error } = await database.rpc("refresh_classification_run_counters", {
    p_run_id: runId,
  });
  assertNoError(error, "Could not refresh classification run counters");
}

export async function requestClassificationCancellation(
  database: DatabaseClient,
  runId: string,
): Promise<void> {
  const { error } = await database
    .from("classification_runs")
    .update({ cancellation_requested_at: new Date().toISOString() })
    .eq("id", runId)
    .in("status", ["queued", "running"]);
  assertNoError(error, "Could not request classification cancellation");
}

export async function getClassificationRun(
  database: DatabaseClient,
  runId: string,
): Promise<ClassificationRunRow> {
  const { data, error } = await database
    .from("classification_runs")
    .select("*")
    .eq("id", runId)
    .single();
  assertNoError(error, "Could not load classification run");
  return data as ClassificationRunRow;
}

export async function listRecentClassificationRuns(
  database: DatabaseClient,
  limit = 20,
): Promise<ClassificationRunRow[]> {
  const { data, error } = await database
    .from("classification_runs")
    .select("*")
    .order("created_at", { ascending: false })
    .limit(limit);
  assertNoError(error, "Could not load classification runs");
  return (data || []) as ClassificationRunRow[];
}

export async function listQueuedClassificationEmailIds(
  database: DatabaseClient,
  runId: string,
): Promise<string[]> {
  const result: string[] = [];
  const pageSize = 1_000;
  for (let offset = 0; ; offset += pageSize) {
    const { data, error } = await database
      .from("email_classifications")
      .select("email_id")
      .eq("run_id", runId)
      .eq("status", "queued")
      .order("created_at")
      .range(offset, offset + pageSize - 1);
    assertNoError(error, "Could not load queued classification emails");
    const rows = (data || []) as Array<{ email_id: string }>;
    result.push(...rows.map((row) => row.email_id));
    if (rows.length < pageSize) return result;
  }
}

export async function startClassificationRun(
  database: DatabaseClient,
  runId: string,
): Promise<void> {
  const current = await getClassificationRun(database, runId);
  if (current.status === "succeeded") {
    throw new Error(`Classification run ${runId} is already terminal (${current.status})`);
  }
  const requeue = await database
    .from("email_classifications")
    .update({ status: "queued", started_at: null })
    .eq("run_id", runId)
    .eq("status", "running");
  assertNoError(requeue.error, "Could not requeue interrupted classifications");
  const { error } = await database
    .from("classification_runs")
    .update({
      status: "running",
      started_at: current.started_at || new Date().toISOString(),
      finished_at: null,
      cancellation_requested_at: null,
      error_message: null,
    })
    .eq("id", runId)
    .in("status", ["queued", "running", "partial", "failed", "cancelled"]);
  assertNoError(error, "Could not start classification run");
}

export async function finishClassificationRun(
  database: DatabaseClient,
  runId: string,
  status: ClassificationRunRow["status"],
  errorMessage: string | null = null,
): Promise<void> {
  const { error } = await database
    .from("classification_runs")
    .update({
      status,
      finished_at: new Date().toISOString(),
      error_message: errorMessage ? sanitizedClassificationError(errorMessage) : null,
    })
    .eq("id", runId);
  assertNoError(error, "Could not finish classification run");
  await refreshClassificationRunCounters(database, runId);
}
